File: airesources/Go/src/hlt/entity.go
package hlt

import (
	"fmt"
	"math"
	"strconv"
)

// DockingStatus represents possible ship.DockingStatus values
type DockingStatus int

const (
	// UNDOCKED ship.DockingStatus value
	UNDOCKED DockingStatus = iota
	// DOCKING ship.DockingStatus value
	DOCKING
	// DOCKED ship.DockingStatus value
	DOCKED
	// UNDOCKING ship.DockingStatus value
	UNDOCKING
)

// Entity captures spacial and ownership state for Planets and Ships
type Entity struct {
	X      float64
	Y      float64
	Radius float64
	Health float64
	Owner  int
	ID     int
}

// Position in 2D space
type Position struct {
	X, Y float64
}

// Planet object from which Halite is mined
type Planet struct {
	Entity
	NumDockingSpots    float64
	NumDockedShips     float64
	CurrentProduction  float64
	RemainingResources float64
	DockedShipIDs      []int
	DockedShips        []Ship
	Owned              float64
	Distance           float64
}

// Ship is a player controlled Entity made for the purpose of doing combat and mining Halite
type Ship struct {
	Entity
	VelX float64
	VelY float64

	PlanetID        int
	Planet          Planet
	DockingStatus   DockingStatus
	DockingProgress float64
	WeaponCooldown  float64
}

// CalculateDistanceTo returns a euclidean distance to the target
func (entity Entity) CalculateDistanceTo(target Entity) float64 {
	dx := target.X - entity.X
	dy := target.Y - entity.Y

	return math.Sqrt(dx*dx + dy*dy)
}

// CalculateAngleTo returns an angle in degrees to the target
func (entity Entity) CalculateAngleTo(target Entity) float64 {
	return RadToDeg(entity.CalculateRadAngleTo(target))
}

// CalculateRadAngleTo returns an angle in radians to the target
func (entity Entity) CalculateRadAngleTo(target Entity) float64 {
	dx := target.X - entity.X
	dy := target.Y - entity.Y

	return math.Atan2(dy, dx)
}

// ClosestPointTo returns the closest point that is at least minDistance from the target
func (entity Entity) ClosestPointTo(target Entity, minDistance float64) Entity {
	dist := entity.CalculateDistanceTo(target) - target.Radius - minDistance
	angle := target.CalculateRadAngleTo(entity)
	x := target.X + dist*math.Cos(angle)
	y := target.Y + dist*math.Sin(angle)
	return Entity{
		X:      x,
		Y:      y,
		Radius: 0,
		Health: 0,
		Owner:  -1,
		ID:     -1,
	}
}

// ParseShip from a slice of game state tokens
func ParseShip(playerID int, tokens []string) (Ship, []string) {
	shipID, _ := strconv.Atoi(tokens[0])
	shipX, _ := strconv.ParseFloat(tokens[1], 64)
	shipY, _ := strconv.ParseFloat(tokens[2], 64)
	shipHealth, _ := strconv.ParseFloat(tokens[3], 64)
	shipVelX, _ := strconv.ParseFloat(tokens[4], 64)
	shipVelY, _ := strconv.ParseFloat(tokens[5], 64)
	shipDockingStatus, _ := strconv.Atoi(tokens[6])
	shipPlanetID, _ := strconv.Atoi(tokens[7])
	shipDockingProgress, _ := strconv.ParseFloat(tokens[8], 64)
	shipWeaponCooldown, _ := strconv.ParseFloat(tokens[9], 64)

	shipEntity := Entity{
		X:      shipX,
		Y:      shipY,
		Radius: .5,
		Health: shipHealth,
		Owner:  playerID,
		ID:     shipID,
	}

	ship := Ship{
		PlanetID:        shipPlanetID,
		DockingStatus:   IntToDockingStatus(shipDockingStatus),
		DockingProgress: shipDockingProgress,
		WeaponCooldown:  shipWeaponCooldown,
		VelX:            shipVelX,
		VelY:            shipVelY,
		Entity:          shipEntity,
	}

	return ship, tokens[10:]
}

// ParsePlanet from a slice of game state tokens
func ParsePlanet(tokens []string) (Planet, []string) {
	planetID, _ := strconv.Atoi(tokens[0])
	planetX, _ := strconv.ParseFloat(tokens[1], 64)
	planetY, _ := strconv.ParseFloat(tokens[2], 64)
	planetHealth, _ := strconv.ParseFloat(tokens[3], 64)
	planetRadius, _ := strconv.ParseFloat(tokens[4], 64)
	planetNumDockingSpots, _ := strconv.ParseFloat(tokens[5], 64)
	planetCurrentProduction, _ := strconv.ParseFloat(tokens[6], 64)
	planetRemainingResources, _ := strconv.ParseFloat(tokens[7], 64)
	planetOwned, _ := strconv.ParseFloat(tokens[8], 64)
	planetOwner, _ := strconv.Atoi(tokens[9])
	planetNumDockedShips, _ := strconv.ParseFloat(tokens[10], 64)

	planetEntity := Entity{
		X:      planetX,
		Y:      planetY,
		Radius: planetRadius,
		Health: planetHealth,
		Owner:  planetOwner,
		ID:     planetID,
	}

	planet := Planet{
		NumDockingSpots:    planetNumDockingSpots,
		NumDockedShips:     planetNumDockedShips,
		CurrentProduction:  planetCurrentProduction,
		RemainingResources: planetRemainingResources,
		DockedShipIDs:      nil,
		DockedShips:        nil,
		Owned:              planetOwned,
		Entity:             planetEntity,
	}

	for i := 0; i < int(planetNumDockedShips); i++ {
		dockedShipID, _ := strconv.Atoi(tokens[11+i])
		planet.DockedShipIDs = append(planet.DockedShipIDs, dockedShipID)
	}
	return planet, tokens[11+int(planetNumDockedShips):]
}

// IntToDockingStatus converts an int to a DockingStatus
func IntToDockingStatus(i int) DockingStatus {
	statuses := [4]DockingStatus{UNDOCKED, DOCKING, DOCKED, UNDOCKING}
	return statuses[i]
}

// Thrust generates a string describing the ship's intension to move during the current turn
func (ship Ship) Thrust(magnitude float64, angle float64) string {
	var boundedAngle int
	if angle > 0.0 {
		boundedAngle = int(math.Floor(angle + .5))
	} else {
		boundedAngle = int(math.Ceil(angle - .5))
	}
	boundedAngle = ((boundedAngle % 360) + 360) % 360
	return fmt.Sprintf("t %s %s %s", strconv.Itoa(ship.ID), strconv.Itoa(int(magnitude)), strconv.Itoa(boundedAngle))
}

// Dock generates a string describing the ship's intension to dock during the current turn
func (ship Ship) Dock(planet Planet) string {
	return fmt.Sprintf("d %s %s", strconv.Itoa(ship.ID), strconv.Itoa(planet.ID))
}

// Undock generates a string describing the ship's intension to undock during the current turn
func (ship Ship) Undock() string {
	return fmt.Sprintf("u %s", strconv.Itoa(ship.ID))
}

// NavigateBasic demonstrates how the player might move ships through space
func (ship Ship) NavigateBasic(target Entity, gameMap Map) string {
	distance := ship.CalculateDistanceTo(target)
	safeDistance := distance - ship.Entity.Radius - target.Radius - .1

	angle := ship.CalculateAngleTo(target)
	speed := 7.0
	if distance < 10 {
		speed = 3.0
	}

	speed = math.Min(speed, safeDistance)
	return ship.Thrust(speed, angle)
}

// CanDock indicates that a ship is close enough to a given planet to dock
func (ship Ship) CanDock(planet Planet) bool {
	dist := ship.CalculateDistanceTo(planet.Entity)

	return dist <= (ship.Radius + planet.Radius + 4)
}

// Navigate demonstrates how the player might negotiate obsticles between
// a ship and its target
func (ship Ship) Navigate(target Entity, gameMap Map) string {
	ob := gameMap.ObstaclesBetween(ship.Entity, target)

	if !ob {
		return ship.NavigateBasic(target, gameMap)
	}

	x0 := math.Min(ship.X, target.X)
	x2 := math.Max(ship.X, target.X)
	y0 := math.Min(ship.Y, target.Y)
	y2 := math.Max(ship.Y, target.Y)

	dx := (x2 - x0) / 5
	dy := (y2 - y0) / 5
	bestdist := 1000.0
	bestTarget := target

	for x1 := x0; x1 <= x2; x1 += dx {
		for y1 := y0; y1 <= y2; y1 += dy {
			intermediateTarget := Entity{
				X:      x1,
				Y:      y1,
				Radius: 0,
				Health: 0,
				Owner:  0,
				ID:     -1,
			}
			ob1 := gameMap.ObstaclesBetween(ship.Entity, intermediateTarget)
			if !ob1 {
				ob2 := gameMap.ObstaclesBetween(intermediateTarget, target)
				if !ob2 {
					totdist := math.Sqrt(math.Pow(x1-x0, 2)+math.Pow(y1-y0, 2)) + math.Sqrt(math.Pow(x1-x2, 2)+math.Pow(y1-y2, 2))
					if totdist < bestdist {
						bestdist = totdist
						bestTarget = intermediateTarget

					}
				}
			}
		}
	}

	return ship.NavigateBasic(bestTarget, gameMap)
}

File: airesources/Haskell/Hlt/Entity.hs
-- | Defines Planets and Ships and various functions that operate on them.
module Hlt.Entity where

import qualified Data.Map as Map
import Text.Read
import Hlt.Constants

-- | Grouping of a Player (bot) and its corresponding Ships.
data Player = Player { playerId :: PlayerId, ships :: Map.Map ShipId Ship } deriving (Show)

-- | An Entity is an object on the GameMap.
class Entity t where
    x :: t -> Float
    y :: t -> Float
    radius :: t -> Float

newtype PlayerId = PlayerId Int
    deriving (Eq, Ord)

instance Show PlayerId where
    show (PlayerId i) = show i

newtype ShipId = ShipId Int
    deriving (Eq, Ord)

instance Show ShipId where
    show (ShipId i) = show i

newtype PlanetId = PlanetId Int
    deriving (Eq, Ord)

instance Show PlanetId where
    show (PlanetId i) = show i

data DockingStatus = Undocked | Docking | Docked | Undocking
     deriving (Eq, Show)

data Location = Location { locationX :: Float
                         , locationY :: Float
                         } deriving (Eq, Show)

data Planet = Planet { planetId :: PlanetId
                     , planetX :: Float
                     , planetY :: Float
                     , planetHealth :: Int
                     , planetRadius :: Float
                     , dockingSpots :: Int
                     , production :: Int
                     , planetOwner :: Maybe PlayerId
                     , dockedShips :: [ShipId]
                     } deriving (Eq, Show)

data Ship = Ship { shipId :: ShipId
                 , shipX :: Float
                 , shipY :: Float
                 , shipHealth :: Int
                 , dockingStatus :: DockingStatus
                 , planet :: Maybe PlanetId
                 , dockingProgress :: Int
                 , weaponCooldown :: Int
                 } deriving (Eq, Show)

instance Entity Location where
    x = locationX
    y = locationY
    radius = const 0

instance Entity Planet where
    x = planetX
    y = planetY
    radius = planetRadius

instance Entity Ship where
    x = shipX
    y = shipY
    radius = pure shipRadius

-- Check if two Entities are equal
equal :: Entity a => Entity b => a -> b -> Bool
equal a b = (x a == x b && y a == y b)

-- Check if two Entities are not equal
notEqual :: Entity a => Entity b => a -> b -> Bool
notEqual a b = not $ equal a b

-- | Returns the x component of the difference between two Entities.
dX :: Entity a => Entity b => a -> b -> Float
dX a b = x b - x a

-- | Returns the y component of the difference between two Entities.
dY :: Entity a => Entity b => a -> b -> Float
dY a b = y b - y a

-- | Returns the angle in radians between two Entities.
angleRadians :: Entity a => Entity b => a -> b -> Float
angleRadians a b = atan2 (dY a b) (dX a b)

-- | Returns the distance between two Entities.
distance :: Entity a => Entity b => a -> b -> Float
distance a b = sqrt ((dX a b)**2 + (dY a b)**2)

-- | Returns the distance between the edges of two Entities.
distanceEdges :: Entity a => Entity b => a -> b -> Float
distanceEdges a b = distance a b - radius a - radius b

-- | Return the closest Location to an Entity
closestLocationTo :: Entity a => Entity b => a -> b -> Location
closestLocationTo s p = do
   let a = angleRadians p s
       r = radius p + 3
   Location ((x p) + r * cos(a)) ((y p) + r * sin(a))

-- | Checks if an Planet is owned by a Player.
isOwned :: Planet -> Bool
isOwned p = planetOwner p /= Nothing

-- | Checks if a Ship is undocked.
isUndocked :: Ship -> Bool
isUndocked s = dockingStatus s == Undocked

-- | Checks if a Ship is docking.
isDocking :: Ship -> Bool
isDocking s = dockingStatus s == Docking

-- | Checks if a Ship is docked.
isDocked :: Ship -> Bool
isDocked s = dockingStatus s == Docked

-- | Checks if a Ship is undocking.
isUndocking :: Ship -> Bool
isUndocking s = dockingStatus s == Undocking

-- | Checks if a Planet is fully docked.
isFull :: Planet -> Bool
isFull p = length (dockedShips p) == dockingSpots p

-- | Checks if a Ship is within docking range of Planet.
canDock :: Ship -> Planet -> Bool
canDock s p = distance s p <= radius p + dockRadius + shipRadius

-- | Constant fudge value so that ships don't collide into to each other.
collisionFudgeFactor :: Float
collisionFudgeFactor = 0.75

-- | Checks whether the line segment between a Ship and an Entity collide with a Circle on the GameMap.
isSegmentCircleCollision :: Entity a => Entity b => Entity c => a -> b -> c -> Bool
isSegmentCircleCollision a b c = do
    let dx = dX a b
        dy = dY a b
        j = dx**2 + dy**2
    if j == 0 then
        (distance a c <= (radius c) + collisionFudgeFactor)
    else do
        let k = -2 * ((x a)**2 - (x a)*(x b) - (x a)*(x c) + (x b)*(x c) + (y a)**2 - (y a)*(y b) - (y a)*(y c) + (y b)*(y c))
            t = min (-k / (2 * j)) 1.0
        ((t >= 0) && ((distance (Location ((x a) + dx * t) ((y a) + dy * t)) c) <= (radius c) + collisionFudgeFactor))
